File: framework/languages/dotnet/src/Zlink.Framework.Contracts/Errors/ZLinkFrameworkException.cs
namespace Zlink.Framework.Contracts.Errors;

public sealed class ZLinkFrameworkException : Exception
{
    internal ZLinkFrameworkException(
        ZLinkFrameworkErrorKind kind,
        string message,
        ZLinkRetryAdvice? retryAdvice = null,
        Exception? innerException = null)
        : base(message, innerException)
    {
        Kind = kind;
        RetryAdvice = retryAdvice ?? DefaultRetryAdvice(kind);
    }

    public ZLinkFrameworkErrorKind Kind { get; }

    internal ZLinkRetryAdvice RetryAdvice { get; }

    /// <summary>
    /// Who produced the failure. <see cref="ZLinkErrorOrigin.Framework"/>
    /// marks errors the framework runtime generated on its own (route
    /// resolution failure, sealed admission, dispatch rejection); their error
    /// replies carry the <c>zlink.origin=framework</c> metadata marker.
    /// <see cref="ZLinkErrorOrigin.Application"/> marks errors decoded from a
    /// peer error reply that did not carry the marker — the remote application
    /// handler produced them, so route caches must not read them as a
    /// stale-route signal. Mirrors the C++ <c>error_origin_t</c> contract.
    /// </summary>
    internal ZLinkErrorOrigin Origin { get; init; }

    private static ZLinkRetryAdvice DefaultRetryAdvice(
        ZLinkFrameworkErrorKind kind) =>
        kind switch
        {
            ZLinkFrameworkErrorKind.Unavailable
                or ZLinkFrameworkErrorKind.CapacityExceeded
                or ZLinkFrameworkErrorKind.DeadlineExceeded =>
                ZLinkRetryAdvice.RetryAfterBackoff,
            ZLinkFrameworkErrorKind.ShuttingDown =>
                ZLinkRetryAdvice.RetryAfterStateChange,
            _ => ZLinkRetryAdvice.DoNotRetry
        };
}

public enum ZLinkFrameworkErrorKind
{
    NotFound = 0,
    AlreadyExists = 1,
    TypeMismatch = 2,
    NotConfigured = 3,
    Rejected = 4,
    Unavailable = 5,
    CapacityExceeded = 6,
    DeadlineExceeded = 7,
    ShuttingDown = 8,
    ProtocolError = 9,
    InvalidOperation = 10,
    DataLost = 11,
    InternalFailure = 12
}

internal enum ZLinkRetryAdvice
{
    DoNotRetry = 0,
    RetryAfterBackoff = 1,
    RetryAfterStateChange = 2
}

internal enum ZLinkErrorOrigin
{
    Unspecified = 0,
    Framework = 1,
    Application = 2
}

File: framework/languages/java/zlink-framework-core/src/main/java/systems/zlink/framework/errors/ZLinkFrameworkException.java
package systems.zlink.framework.errors;

import java.util.Map;

public class ZLinkFrameworkException extends RuntimeException {
    private final ZLinkFrameworkErrorKind kind;
    private final Map<String, String> metadata;

    public ZLinkFrameworkException(ZLinkFrameworkErrorKind kind, String message) {
        this(kind, message, null);
    }

    public ZLinkFrameworkException(ZLinkFrameworkErrorKind kind, String message, Throwable cause) {
        this(kind, message, cause, Map.of());
    }

    public ZLinkFrameworkException(
        ZLinkFrameworkErrorKind kind,
        String message,
        Throwable cause,
        Map<String, String> metadata) {
        super(message, cause);
        this.kind = kind == null
            ? ZLinkFrameworkErrorKind.INTERNAL_FAILURE
            : kind;
        this.metadata = metadata == null || metadata.isEmpty()
            ? Map.of()
            : Map.copyOf(metadata);
    }

    public ZLinkFrameworkErrorKind kind() {
        return kind;
    }

    /**
     * Framework-owned reply metadata carried with the error. Framework-generated
     * errors mark their origin here ({@code zlink.origin=framework}); errors an
     * application handler raised never carry that marker.
     */
    public Map<String, String> metadata() {
        return metadata;
    }

}
